nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarConsumerProcessor.java

```diff
@@ -0,0 +1,296 @@
+                    + "in pipeline is removed")
+            .required(false)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .defaultValue("1000")
+            .build();
+
+    public static final PropertyDescriptor SUBSCRIPTION_TYPE = new PropertyDescriptor.Builder()
+            .name("Subscription Type")
+            .description("Select the subscription type to be used when subscribing to the topic.")
+            .required(false)
+            .allowableValues(EXCLUSIVE, SHARED, FAILOVER)
+            .defaultValue(SHARED.getValue())
+            .build();
+
+    protected static final List<PropertyDescriptor> PROPERTIES;
+    protected static final Set<Relationship> RELATIONSHIPS;
+
+    static {
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.add(PULSAR_CLIENT_SERVICE);
+        properties.add(TOPICS);
+        properties.add(TOPICS_PATTERN);
+        properties.add(SUBSCRIPTION_NAME);
+        properties.add(CONSUMER_NAME);
+        properties.add(ASYNC_ENABLED);
+        properties.add(MAX_ASYNC_REQUESTS);
+        properties.add(ACK_TIMEOUT);
+        properties.add(PRIORITY_LEVEL);
+        properties.add(RECEIVER_QUEUE_SIZE);
+        properties.add(SUBSCRIPTION_TYPE);
+
+        PROPERTIES = Collections.unmodifiableList(properties);
+
+        final Set<Relationship> relationships = new HashSet<>();
+        relationships.add(REL_SUCCESS);
+        RELATIONSHIPS = Collections.unmodifiableSet(relationships);
+    }
+
+    protected Consumer<T> consumer;
+    protected ExecutorService consumerPool;
+    protected ExecutorCompletionService<Message<T>> consumerService;
+    protected ExecutorService ackPool;
+    protected ExecutorCompletionService<Object> ackService;
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @OnScheduled
+    public void init(ProcessContext context) {
+       if (context.getProperty(ASYNC_ENABLED).isSet() && context.getProperty(ASYNC_ENABLED).asBoolean()) {
+           consumerPool = Executors.newFixedThreadPool(context.getProperty(MAX_ASYNC_REQUESTS).asInteger());
+           consumerService = new ExecutorCompletionService<>(consumerPool);
+           ackPool = Executors.newFixedThreadPool(context.getProperty(MAX_ASYNC_REQUESTS).asInteger() + 1);
+           ackService = new ExecutorCompletionService<>(ackPool);
+       }
+    }
+
+    @OnUnscheduled
+    public void shutDown(final ProcessContext context) {
+        if (context.getProperty(ASYNC_ENABLED).isSet() && context.getProperty(ASYNC_ENABLED).asBoolean()) {
+            try {
+                consumerPool.shutdown();
+                consumerPool.awaitTermination(10, TimeUnit.SECONDS);
```

Comment:
I added the wait after some of my unit tests showed that not all of the messages consumed where being acknowledged. Adding the delay provides some additional opportunity for the acknowledgements to occur.  By ensuring that the messages are properly acked, we prevent re-processing the same messages in the event of a shutdown and restart of the processor since the un-acked messages would be replayed on startup.

Reply:
OK, that makes sense. Thanks for the clarification. It probably makes sense to leave an inline comment there, then, to explain the reasoning behind it.